src/main/scala/com/databricks/labs/overwatch/pipeline/GoldTransforms.scala

```diff
@@ -493,6 +493,67 @@ trait GoldTransforms extends SparkSessionWrapper {
 
   }
 
+
+  protected def buildNotebookCommands(
+                                   notebook: PipelineTable,
+                                   clsfIncrementalDF : DataFrame,
+                                 )(auditIncrementalDF: DataFrame): DataFrame = {
+
+
+    val notebookLookupTSDF = notebook.asDF
+      .select("organization_id", "notebook_id", "notebook_path", "notebook_name", "unixTimeMS","date")
+      .withColumnRenamed("notebook_id", "notebookId")
+      .distinct
+      .toTSDF("unixTimeMS", "organization_id","date","notebookId")
+
+    val clsfLookupTSDF = clsfIncrementalDF
+      .select(
+        "organization_id", "state_start_date", "unixTimeMS_state_start", "cluster_id", "cluster_name",
+        "custom_tags", "node_type_id", "current_num_workers", "uptime_in_state_H", "total_cost")
+      .distinct
+      .withColumnRenamed("unixTimeMS_state_start", "unixTimeMS")
+      .withColumnRenamed("cluster_id", "clusterId")
+      .withColumnRenamed("current_num_workers", "node_count")
+      .withColumn("totalCostPMS", col("total_cost") / col("uptime_in_state_H"))
+      .toTSDF("unixTimeMS", "organization_id", "clusterId")
+
+    val colNames = Array(
+      "organization_id", "cluster_id","notebook_id", "workspace_name", "date", "timestamp",
+       "notebook_path", "notebook_name", "command_id", "command_text", "execution_time_s", "source_ip_address",
+      "user_identity", "estimated_dbu_cost", "status", "cluster_name", "custom_tags",
+      "node_type_id", "node_count", "response", "user_agent", "unixTimeMS"
+    )
+
+    val notebookCodeAndMetaDF = auditIncrementalDF
+      .filter(col("serviceName") === "notebook" && col("actionName") === "runCommand")
+      .verifyMinimumSchema(Schema.auditMasterSchema)
+      .selectExpr("*", "requestParams.*").drop("requestParams")
+      .withColumnRenamed("timestamp", "unixTimeMS")
+      .withColumn("timestamp", from_unixtime(col("unixTimeMS") / 1000.0).cast("timestamp"))
+      .withColumn("executionTime",col("executionTime").cast("double"))
+      .toTSDF("unixTimeMS", "organization_id", "date","notebookId")
+      .lookupWhen(notebookLookupTSDF)
```

Comment:
you might get more notebook_paths / meta if you allow a lookAhead here -- suggest setting lookAhead to like 100 rows or something -- lookback can probably remain max since we shouldn't ever have an insane number of records in a single run.

Reply:
Done